File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/controllers/Keyfob.kt
package com.mentra.bluetoothsdk.controllers

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.core.content.ContextCompat
import com.mentra.bluetoothsdk.Bridge
import com.mentra.bluetoothsdk.DeviceStore
import com.mentra.bluetoothsdk.PhotoRequest
import com.mentra.bluetoothsdk.controllers.keyfob.KeyfobProtocol
import com.mentra.bluetoothsdk.utils.ControllerTypes
import java.util.ArrayDeque
import java.util.UUID

/**
 * Unofficial MentraOS controller for Seeed Studio XIAO nRF52840 Plus.
 * Not a Seeed product and not affiliated with Seeed Studio.
 */
class Keyfob : ControllerManager() {
    companion object {
        private const val TAG = "Keyfob"
        private const val SCAN_DURATION_MS = 15_000L
        private const val PREFS = "keyfob_prefs"
        private const val PREF_BLE_ADDRESS = "keyfob_bleAddress"
        private const val PREF_NAME = "keyfob_name"

        @JvmStatic
        fun matchesAdvertisedName(name: String?): Boolean = KeyfobProtocol.matchesAdvertisedName(name)
    }

    private val appContext: Context = Bridge.getContext()
    private val prefs = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    private val handler = Handler(Looper.getMainLooper())
    private var bluetoothAdapter: BluetoothAdapter? = BluetoothAdapter.getDefaultAdapter()

    private var scanner: BluetoothLeScanner? = null
    private var scanCallback: ScanCallback? = null
    private var scanTimeout: Runnable? = null
    private var deviceSearchId: String = "NOT_SET"
    private val discoveredNames = HashSet<String>()

    private var gatt: BluetoothGatt? = null
    private var ctrlChar: BluetoothGattCharacteristic? = null
    private var evtChar: BluetoothGattCharacteristic? = null
    private val notifyQueue: ArrayDeque<BluetoothGattCharacteristic> = ArrayDeque()
    private var notifyWriteInFlight = false
    private val writeQueue: ArrayDeque<ByteArray> = ArrayDeque()
    private var writeInFlight = false
    private var seq = 1
    private var isDisconnecting = false

    private var bleAddress: String?
        get() = prefs.getString(PREF_BLE_ADDRESS, null)
        set(value) {
            prefs.edit().apply {
                if (value == null) remove(PREF_BLE_ADDRESS) else putString(PREF_BLE_ADDRESS, value)
                apply()
            }
        }

    private var savedName: String?
        get() = prefs.getString(PREF_NAME, null)
        set(value) {
            prefs.edit().apply {
                if (value == null) remove(PREF_NAME) else putString(PREF_NAME, value)
                apply()
            }
        }

    init {
        type = ControllerTypes.KEYFOB
        hasMic = false
    }

    private fun hasScanPermission(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return true
        return ContextCompat.checkSelfPermission(
            appContext, Manifest.permission.BLUETOOTH_SCAN
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun hasConnectPermission(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return true
        return ContextCompat.checkSelfPermission(
            appContext, Manifest.permission.BLUETOOTH_CONNECT
        ) == PackageManager.PERMISSION_GRANTED
    }

    override fun findCompatibleDevices() {
        Bridge.log("$TAG: findCompatibleDevices()")
        deviceSearchId = "NOT_SET"
        discoveredNames.clear()
        startScan(forConnection = false)
    }

    override fun connectById(id: String) {
        Bridge.log("$TAG: connectById($id)")
        deviceSearchId = id
        if (connectByStoredAddress(id)) return
        startScan(forConnection = true)
    }

    override fun stopScan() {
        scanTimeout?.let { handler.removeCallbacks(it) }
        scanTimeout = null
        val s = scanner ?: return
        val cb = scanCallback ?: return
        try {
            s.stopScan(cb)
        } catch (e: SecurityException) {
            Bridge.log("$TAG: stopScan SecurityException: ${e.message}")
        }
        scanCallback = null
        scanner = null
    }

    private fun startScan(forConnection: Boolean) {
        Bridge.log("$TAG: startScan(forConnection=$forConnection)")
        val adapter = bluetoothAdapter ?: BluetoothAdapter.getDefaultAdapter().also { bluetoothAdapter = it }
        if (adapter == null || !adapter.isEnabled) {
            Bridge.log("$TAG: Bluetooth unavailable or disabled")
            return
        }
        if (!hasScanPermission() || !hasConnectPermission()) {
            Bridge.log("$TAG: Missing Bluetooth permissions")
            return
        }
        if (gatt != null) {
            Bridge.log("$TAG: Already connected, skipping scan")
            return
        }
        isDisconnecting = false
        stopScan()
        scanner = adapter.bluetoothLeScanner
        if (scanner == null) {
            Bridge.log("$TAG: BLE scanner unavailable")
            return
        }
        scanCallback =
            object : ScanCallback() {
                override fun onScanResult(callbackType: Int, result: ScanResult) {
                    handleScanResult(result, forConnection)
                }

                override fun onScanFailed(errorCode: Int) {
                    Bridge.log("$TAG: scan failed code=$errorCode")
                    stopScan()
                }
            }
        val settings = ScanSettings.Builder().setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY).build()
        try {
            scanner?.startScan(emptyList(), settings, scanCallback)
        } catch (e: SecurityException) {
            Bridge.log("$TAG: startScan SecurityException: ${e.message}")
            return
        }
        scanTimeout = Runnable { stopScan() }
        handler.postDelayed(scanTimeout!!, SCAN_DURATION_MS)
    }

    private fun advertisedName(result: ScanResult): String? {
        val fromRecord = result.scanRecord?.deviceName
        return try {
            result.device?.name ?: fromRecord
        } catch (e: SecurityException) {
            fromRecord
        }
    }

    private fun handleScanResult(result: ScanResult, forConnection: Boolean) {
        val device = result.device ?: return
        val advertised = advertisedName(result)
        if (!matchesAdvertisedName(advertised)) return
        val name = advertised ?: return
        val address = device.address.orEmpty()
        handler.post {
            if (!forConnection) {
                if (discoveredNames.add(name)) {
                    Bridge.sendDiscoveredDevice(type, name, address, result.rssi)
                }
                return@post
            }
            if (gatt != null) {
                stopScan()
                return@post
            }
            val target = deviceSearchId
            val matchesTarget =
                target == "NOT_SET" ||
                    target.isBlank() ||
                    target.equals(address, ignoreCase = true) ||
                    target.equals(name, ignoreCase = true) ||
                    name.contains(target, ignoreCase = true)
            if (!matchesTarget) return@post
            stopScan()
            connectGatt(device)
        }
    }

    private fun connectByStoredAddress(id: String): Boolean {
        val address = bleAddress ?: return false
        val knownName = savedName
        val idMatches =
            id.isBlank() ||
                id == "NOT_SET" ||
                id.equals(address, ignoreCase = true) ||
                (!knownName.isNullOrBlank() && id.equals(knownName, ignoreCase = true))
        if (!idMatches) return false
        val adapter = bluetoothAdapter ?: return false
        val device = try {
            adapter.getRemoteDevice(address)
        } catch (e: IllegalArgumentException) {
            Bridge.log("$TAG: Invalid stored BLE address: $address")
            return false
        }
        connectGatt(device)
        return true
    }

    private fun connectGatt(device: BluetoothDevice) {
        cleanupGatt(keepState = true)
        try {
            bleAddress = device.address
            gatt =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    device.connectGatt(appContext, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
                } else {
                    device.connectGatt(appContext, false, gattCallback)
                }
            Bridge.log("$TAG: Connecting to ${device.name ?: device.address}")
        } catch (e: SecurityException) {
            Bridge.log("$TAG: connectGatt SecurityException: ${e.message}")
        }
    }

    private val gattCallback =
        object : BluetoothGattCallback() {
            override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
                if (newState == BluetoothProfile.STATE_CONNECTED) {
                    Bridge.log("$TAG: connected to ${gatt.device?.name}")
                    this@Keyfob.gatt = gatt
                    handler.postDelayed({
                        try {
                            gatt.discoverServices()
                        } catch (e: SecurityException) {
                            Bridge.log("$TAG: discoverServices SecurityException: ${e.message}")
                        }
                    }, 200)
                } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                    Bridge.log("$TAG: disconnected status=$status")
                    resetControllerState()
                    cleanupGatt()
                }
            }

            override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
                if (status != BluetoothGatt.GATT_SUCCESS) {
                    Bridge.log("$TAG: service discovery failed status=$status")
                    return
                }
                val service = gatt.getService(KeyfobProtocol.SERVICE_UUID)
                if (service == null) {
                    Bridge.log("$TAG: Keyfob GATT service missing")
                    return
                }
                ctrlChar = service.getCharacteristic(KeyfobProtocol.CTRL_UUID)
                evtChar = service.getCharacteristic(KeyfobProtocol.EVT_UUID)
                notifyQueue.clear()
                evtChar?.let { notifyQueue.add(it) }
                if (notifyQueue.isEmpty()) {
                    markReady()
                } else {
                    processNextNotify()
                }
            }

            override fun onDescriptorWrite(
                gatt: BluetoothGatt,
                descriptor: BluetoothGattDescriptor,
                status: Int,
            ) {
                notifyWriteInFlight = false
                if (status != BluetoothGatt.GATT_SUCCESS) {
                    Bridge.log("$TAG: descriptor write failed status=$status")
                }
                processNextNotify()
            }

            override fun onCharacteristicWrite(
                gatt: BluetoothGatt,
                characteristic: BluetoothGattCharacteristic,
                status: Int,
            ) {
                writeInFlight = false
                if (status != BluetoothGatt.GATT_SUCCESS) {
                    Bridge.log("$TAG: characteristic write failed status=$status")
                }
                pumpWrites()
            }

            override fun onCharacteristicChanged(
                gatt: BluetoothGatt,
                characteristic: BluetoothGattCharacteristic,
            ) {
                handleNotify(characteristic.uuid, characteristic.value)
            }

            override fun onCharacteristicChanged(
                gatt: BluetoothGatt,
                characteristic: BluetoothGattCharacteristic,
                value: ByteArray,
            ) {
                handleNotify(characteristic.uuid, value)
            }
        }

    private fun processNextNotify() {
        val localGatt = gatt ?: return
        if (notifyWriteInFlight) return
        val next = notifyQueue.poll()
        if (next == null) {
            markReady()
            return
        }
        try {
            localGatt.setCharacteristicNotification(next, true)
        } catch (e: SecurityException) {
            Bridge.log("$TAG: setCharacteristicNotification SecurityException: ${e.message}")
            processNextNotify()
            return
        }
        val cccd = next.getDescriptor(KeyfobProtocol.CCCD_UUID)
        if (cccd == null) {
            processNextNotify()
            return
        }
        notifyWriteInFlight = true
        cccd.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
        try {
            localGatt.writeDescriptor(cccd)
        } catch (e: SecurityException) {
            notifyWriteInFlight = false
            Bridge.log("$TAG: writeDescriptor SecurityException: ${e.message}")
            processNextNotify()
        }
    }

    private fun markReady() {
        val connectedName = getConnectedBluetoothName()
        if (!connectedName.isNullOrBlank()) {
            savedName = connectedName
            DeviceStore.apply("bluetooth", "controller_device_name", connectedName)
        }
        val address = try { gatt?.device?.address } catch (e: SecurityException) { null }
        if (!address.isNullOrBlank()) {
            bleAddress = address
            DeviceStore.apply("glasses", "controllerMacAddress", address)
        }
        DeviceStore.apply("glasses", "controllerConnected", true)
        DeviceStore.apply("glasses", "controllerFullyBooted", true)
        Bridge.log("$TAG: ready")
    }

    private fun handleNotify(uuid: UUID?, value: ByteArray?) {
        if (uuid != KeyfobProtocol.EVT_UUID) return
        val decoded = KeyfobProtocol.decode(value) ?: return
        val (opcode, _, payload) = decoded
        when (opcode) {
            KeyfobProtocol.EVT_BATTERY -> {
                if (payload.isNotEmpty()) {
                    val percent = (payload[0].toInt() and 0xFF).coerceIn(0, 100)
                    DeviceStore.apply("glasses", "controllerBatteryLevel", percent)
                }
            }
            KeyfobProtocol.EVT_READY -> markReady()
            KeyfobProtocol.EVT_GESTURE -> {
                if (payload.isEmpty()) return
                val name = KeyfobProtocol.gestureName(payload[0]) ?: return
                val source = if (payload.size > 1) payload[1].toInt() and 0xFF else null
                Bridge.log("$TAG: Gesture: $name button=$source")
                Bridge.sendTouchEvent(type, name, System.currentTimeMillis(), source)
            }
            else -> {}
        }
    }

    private fun enqueueControl(opcode: Byte, payload: ByteArray = ByteArray(0)) {
        writeQueue.add(KeyfobProtocol.encode(opcode, nextSeq(), payload))
        pumpWrites()
    }

    private fun nextSeq(): Int {
        val current = seq
        seq = if (seq == 255) 1 else seq + 1
        return current
    }

    private fun pumpWrites() {
        val localGatt = gatt ?: return
        val characteristic = ctrlChar ?: return
        if (writeInFlight) return
        val packet = writeQueue.poll() ?: return
        writeInFlight = true
        characteristic.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
        characteristic.value = packet
        try {
            localGatt.writeCharacteristic(characteristic)
        } catch (e: SecurityException) {
            writeInFlight = false
            Bridge.log("$TAG: writeCharacteristic SecurityException: ${e.message}")
        }
    }

    private fun resetControllerState() {
        DeviceStore.apply("glasses", "controllerConnected", false)
        DeviceStore.apply("glasses", "controllerFullyBooted", false)
    }

    private fun cleanupGatt(keepState: Boolean = false) {
        writeQueue.clear()
        writeInFlight = false
        notifyQueue.clear()
        notifyWriteInFlight = false
        ctrlChar = null
        evtChar = null
        try {
            gatt?.disconnect()
            gatt?.close()
        } catch (e: SecurityException) {
            Bridge.log("$TAG: cleanupGatt SecurityException: ${e.message}")
        }
        gatt = null
        if (!keepState && !isDisconnecting) {
            resetControllerState()
        }
    }

    override fun disconnect() {
        Bridge.log("$TAG: disconnect()")
        isDisconnecting = true
        stopScan()
        resetControllerState()
        cleanupGatt(keepState = true)
    }

    override fun forget() {
        disconnect()
        bleAddress = null
        savedName = null
        deviceSearchId = "NOT_SET"
    }

    override fun cleanup() {
        disconnect()
    }

    override fun getConnectedBluetoothName(): String? {
        return try {
            gatt?.device?.name
        } catch (e: SecurityException) {
            null
        }
    }

    override fun ping() {
        enqueueControl(KeyfobProtocol.CMD_PING)
    }

    override fun getBatteryStatus() {
        enqueueControl(KeyfobProtocol.CMD_PING)
    }

    override fun sendRgbLedControl(
        requestId: String,
        packageName: String?,
        action: String,
        color: String?,
        onDurationMs: Int,
        offDurationMs: Int,
        count: Int,
    ) {
        val rgb =
            if (action.equals("off", ignoreCase = true)) {
                byteArrayOf(0, 0, 0)
            } else {
                parseRgb(color)
            }
        enqueueControl(KeyfobProtocol.CMD_LED, rgb)
    }

    private fun parseRgb(color: String?): ByteArray {
        val raw = color?.trim().orEmpty()
        if (raw.startsWith("#") && raw.length >= 7) {
            val r = raw.substring(1, 3).toIntOrNull(16) ?: 0
            val g = raw.substring(3, 5).toIntOrNull(16) ?: 0
            val b = raw.substring(5, 7).toIntOrNull(16) ?: 0
            return byteArrayOf(r.toByte(), g.toByte(), b.toByte())
        }
        return when (raw.lowercase()) {
            "red" -> byteArrayOf(255.toByte(), 0, 0)
            "green" -> byteArrayOf(0, 255.toByte(), 0)
            "blue" -> byteArrayOf(0, 0, 255.toByte())
            "white" -> byteArrayOf(255.toByte(), 255.toByte(), 255.toByte())
            else -> byteArrayOf(255.toByte(), 255.toByte(), 255.toByte())
        }
    }

    override fun sendIncidentId(incidentId: String) {}
    override fun setMicEnabled(enabled: Boolean) {}
    override fun sortMicRanking(list: MutableList<String>): MutableList<String> = list
    override fun sendJson(jsonOriginal: Map<String, Any>, wakeUp: Boolean, requireAck: Boolean) {}
    override fun requestPhoto(request: PhotoRequest) {}
    override fun startVideoRecording(requestId: String, save: Boolean, sound: Boolean) {}
    override fun stopVideoRecording(requestId: String) {}
    override fun startStream(message: Map<String, Any>) {}
    override fun stopStream() {}
    override fun sendStreamKeepAlive(message: Map<String, Any>) {}
    override fun sendButtonPhotoSettings() {}
    override fun sendButtonVideoRecordingSettings() {}
    override fun sendButtonMaxRecordingTime() {}
    override fun setBrightness(level: Int, autoMode: Boolean) {}
    override fun clearDisplay() {}
    override fun sendTextWall(text: String) {}
    override fun sendDoubleTextWall(top: String, bottom: String) {}
    override fun displayBitmap(
        base64ImageData: String,
        x: Int?,
        y: Int?,
        width: Int?,
        height: Int?,
    ): Boolean = false
    override fun showDashboard() {}
    override fun setDashboardPosition(height: Int, depth: Int) {}
    override fun setHeadUpAngle(angle: Int) {}
    override fun setSilentMode(enabled: Boolean) {}
    override fun exit() {}
    override fun sendShutdown() {
        disconnect()
    }
    override fun sendReboot() {}
    override fun requestWifiScan() {}
    override fun sendWifiCredentials(ssid: String, password: String) {}
    override fun forgetWifiNetwork(ssid: String) {}
    override fun sendHotspotState(enabled: Boolean) {}
    override fun sendOtaStart(otaVersionUrl: String?) {}
    override fun sendUserEmailToGlasses(email: String) {}
    override fun queryGalleryStatus() {}
    override fun sendGalleryMode() {}
    override fun requestVersionInfo() {}
}

File: mobile/modules/bluetooth-sdk/ios/Source/controllers/Keyfob.swift
//
//  Keyfob.swift
//  MentraOS_Manager
//
//  Unofficial MentraOS controller for Seeed Studio XIAO nRF52840 Plus.
//  Not a Seeed product; not affiliated with or endorsed by Seeed Studio.
//  Protocol shared with firmware/keyfob/settings.h and KeyfobProtocol.kt.
//

import CoreBluetooth
import Foundation

private enum KeyfobBLE {
    static let SERVICE_UUID = CBUUID(string: "d4b2c520-8f1e-4c7a-9b03-6a5d4e80e020")
    static let CTRL_UUID = CBUUID(string: "d4b2c520-8f1e-4c7a-9b03-6a5d4e80e021")
    static let EVT_UUID = CBUUID(string: "d4b2c520-8f1e-4c7a-9b03-6a5d4e80e022")
    static let NAME_PREFIX = "Keyfob"
    static let HDR_LEN = 4
    static let CMD_LED: UInt8 = 0x01
    static let CMD_PING: UInt8 = 0x02
    static let EVT_BATTERY: UInt8 = 0x81
    static let EVT_READY: UInt8 = 0x82
    static let EVT_GESTURE: UInt8 = 0x83
    static let SCAN_TIMEOUT: TimeInterval = 15.0

    static func matchesName(_ name: String?) -> Bool {
        guard let name, !name.trimmingCharacters(in: .whitespaces).isEmpty else { return false }
        return name.lowercased().hasPrefix(NAME_PREFIX.lowercased())
    }

    static func gestureName(_ id: UInt8) -> String? {
        switch id {
        case 0x01: return "hold"
        case 0x02: return "single_tap"
        case 0x03: return "double_tap"
        case 0x04: return "swipe_up"
        case 0x05: return "swipe_down"
        default: return nil
        }
    }

    static func encode(opcode: UInt8, seq: UInt8, payload: Data = Data()) -> Data {
        var out = Data(count: HDR_LEN + payload.count)
        out[0] = opcode
        out[1] = seq
        out[2] = UInt8(payload.count & 0xFF)
        out[3] = UInt8((payload.count >> 8) & 0xFF)
        if !payload.isEmpty {
            out.replaceSubrange(HDR_LEN ..< (HDR_LEN + payload.count), with: payload)
        }
        return out
    }

    static func decode(_ packet: Data?) -> (UInt8, UInt8, Data)? {
        guard let packet, packet.count >= HDR_LEN else { return nil }
        let len = Int(packet[2]) | (Int(packet[3]) << 8)
        let end = min(HDR_LEN + len, packet.count)
        return (packet[0], packet[1], packet.subdata(in: HDR_LEN ..< end))
    }
}

@MainActor
class Keyfob: NSObject, ControllerManager {
    var type = ControllerTypes.KEYFOB
    let hasMic = false

    private var centralManager: CBCentralManager?
    private var peripheral: CBPeripheral?
    private var ctrlChar: CBCharacteristic?
    private var evtChar: CBCharacteristic?
    private var isDisconnecting = false
    private var seq: UInt8 = 1
    var DEVICE_SEARCH_ID = "NOT_SET"
    private var discoveredNames = Set<String>()
    private var scanTimeoutWork: DispatchWorkItem?

    private var savedUUID: UUID? {
        get { UserDefaults.standard.string(forKey: "keyfob_uuid").flatMap { UUID(uuidString: $0) } }
        set {
            if let v = newValue {
                UserDefaults.standard.set(v.uuidString, forKey: "keyfob_uuid")
            } else {
                UserDefaults.standard.removeObject(forKey: "keyfob_uuid")
            }
        }
    }

    private var savedName: String? {
        get { UserDefaults.standard.string(forKey: "keyfob_name") }
        set { UserDefaults.standard.set(newValue, forKey: "keyfob_name") }
    }

    static let bluetoothQueue = DispatchQueue(label: "BluetoothKeyfob", qos: .userInitiated)

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: Keyfob.bluetoothQueue)
    }

    func findCompatibleDevices() {
        Bridge.log("Keyfob: findCompatibleDevices()")
        DEVICE_SEARCH_ID = "NOT_SET"
        discoveredNames.removeAll()
        startScan()
    }

    func connectById(_ id: String) {
        Bridge.log("Keyfob: connectById(\(id))")
        DEVICE_SEARCH_ID = id
        if connectBySavedUUID(id) { return }
        startScan()
    }

    func stopScan() {
        scanTimeoutWork?.cancel()
        scanTimeoutWork = nil
        centralManager?.stopScan()
    }

    private func startScan() {
        guard let centralManager else { return }
        if centralManager.state != .poweredOn {
            Bridge.log("Keyfob: Bluetooth not powered on")
            return
        }
        if peripheral != nil {
            Bridge.log("Keyfob: Already connected, skipping scan")
            return
        }
        isDisconnecting = false
        stopScan()
        centralManager.scanForPeripherals(withServices: nil, options: [CBCentralManagerScanOptionAllowDuplicatesKey: false])
        let work = DispatchWorkItem { [weak self] in
            self?.stopScan()
        }
        scanTimeoutWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + KeyfobBLE.SCAN_TIMEOUT, execute: work)
    }

    private func connectBySavedUUID(_ id: String) -> Bool {
        guard let uuid = savedUUID, let centralManager else { return false }
        let known = savedName
        let matches =
            id.isEmpty || id == "NOT_SET" || uuid.uuidString.caseInsensitiveCompare(id) == .orderedSame
            || (known?.caseInsensitiveCompare(id) == .orderedSame)
        guard matches else { return false }
        let retrieved = centralManager.retrievePeripherals(withIdentifiers: [uuid])
        guard let target = retrieved.first else { return false }
        connect(target)
        return true
    }

    private func connect(_ peripheral: CBPeripheral) {
        stopScan()
        self.peripheral = peripheral
        peripheral.delegate = self
        savedUUID = peripheral.identifier
        centralManager?.connect(peripheral, options: nil)
        Bridge.log("Keyfob: Connecting to \(peripheral.name ?? peripheral.identifier.uuidString)")
    }

    private func nextSeq() -> UInt8 {
        let current = seq
        seq = seq == 255 ? 1 : seq &+ 1
        return current
    }

    private func enqueueControl(opcode: UInt8, payload: Data = Data()) {
        guard let ctrlChar, let peripheral else { return }
        let packet = KeyfobBLE.encode(opcode: opcode, seq: nextSeq(), payload: payload)
        peripheral.writeValue(packet, for: ctrlChar, type: .withResponse)
    }

    private func markReady() {
        if let name = peripheral?.name, !name.isEmpty {
            savedName = name
            DeviceStore.shared.apply("bluetooth", "controller_device_name", name)
        }
        if let uuid = peripheral?.identifier {
            savedUUID = uuid
            DeviceStore.shared.apply("glasses", "controllerMacAddress", uuid.uuidString)
        }
        DeviceStore.shared.apply("glasses", "controllerConnected", true)
        DeviceStore.shared.apply("glasses", "controllerFullyBooted", true)
        Bridge.log("Keyfob: ready")
    }

    private func resetControllerState() {
        DeviceStore.shared.apply("glasses", "controllerConnected", false)
        DeviceStore.shared.apply("glasses", "controllerFullyBooted", false)
    }

    func disconnect() {
        Bridge.log("Keyfob: disconnect()")
        isDisconnecting = true
        stopScan()
        resetControllerState()
        if let peripheral {
            centralManager?.cancelPeripheralConnection(peripheral)
        }
        self.peripheral = nil
        ctrlChar = nil
        evtChar = nil
    }

    func forget() {
        disconnect()
        savedUUID = nil
        savedName = nil
        DEVICE_SEARCH_ID = "NOT_SET"
    }

    func cleanup() {
        disconnect()
    }

    func getConnectedBluetoothName() -> String? {
        peripheral?.name
    }

    func ping() {
        enqueueControl(opcode: KeyfobBLE.CMD_PING)
    }

    func getBatteryStatus() {
        enqueueControl(opcode: KeyfobBLE.CMD_PING)
    }

    func sendRgbLedControl(
        requestId _: String, packageName _: String?, action: String, color: String?, onDurationMs _: Int,
        offDurationMs _: Int, count _: Int
    ) {
        let rgb: Data
        if action.lowercased() == "off" {
            rgb = Data([0, 0, 0])
        } else {
            rgb = parseRgb(color)
        }
        enqueueControl(opcode: KeyfobBLE.CMD_LED, payload: rgb)
    }

    private func parseRgb(_ color: String?) -> Data {
        let raw = color?.trimmingCharacters(in: .whitespaces) ?? ""
        if raw.hasPrefix("#"), raw.count >= 7 {
            let r = UInt8(raw.dropFirst().prefix(2), radix: 16) ?? 0
            let g = UInt8(raw.dropFirst(3).prefix(2), radix: 16) ?? 0
            let b = UInt8(raw.dropFirst(5).prefix(2), radix: 16) ?? 0
            return Data([r, g, b])
        }
        switch raw.lowercased() {
        case "red": return Data([255, 0, 0])
        case "green": return Data([0, 255, 0])
        case "blue": return Data([0, 0, 255])
        default: return Data([255, 255, 255])
        }
    }

    func sendIncidentId(_: String, apiBaseUrl _: String?) {}
    func setMicEnabled(_: Bool) {}
    func sortMicRanking(list: [String]) -> [String] { list }
    func sendJson(_: [String: Any], wakeUp _: Bool, requireAck _: Bool) {}
    func requestPhoto(_: PhotoRequest) {}
    func startVideoRecording(requestId _: String, save _: Bool, sound _: Bool) {}
    func stopVideoRecording(requestId _: String) {}
    func startStream(_: [String: Any]) {}
    func stopStream() {}
    func sendStreamKeepAlive(_: [String: Any]) {}
    func sendButtonPhotoSettings() {}
    func sendButtonVideoRecordingSettings() {}
    func sendButtonMaxRecordingTime() {}
    func setBrightness(_: Int, autoMode _: Bool) {}
    func clearDisplay() {}
    func sendTextWall(_: String) {}
    func sendDoubleTextWall(_: String, _: String) {}
    func displayBitmap(
        base64ImageData _: String, x _: Int32? = nil, y _: Int32? = nil, width _: Int32? = nil, height _: Int32? = nil
    ) async -> Bool {
        false
    }
    func showDashboard() {}
    func setDashboardPosition(_: Int, _: Int) {}
    func setHeadUpAngle(_: Int) {}
    func setSilentMode(_: Bool) {}
    func exit() {}
    func sendShutdown() { disconnect() }
    func sendReboot() {}
    func requestWifiScan() {}
    func sendWifiCredentials(_: String, _: String) {}
    func forgetWifiNetwork(_: String) {}
    func sendHotspotState(_: Bool) {}
    func sendOtaStart(otaVersionUrl _: String?) {}
    func sendOtaQueryStatus() {}
    func sendUserEmailToGlasses(_: String) {}
    func queryGalleryStatus() {}
    func sendGalleryMode() {}
    func requestVersionInfo() {}
}

extension Keyfob: CBCentralManagerDelegate {
    nonisolated func centralManagerDidUpdateState(_ central: CBCentralManager) {
        DispatchQueue.main.async { [weak self] in
            Bridge.log("Keyfob: Bluetooth state: \(central.state.rawValue)")
            if central.state == .poweredOn {
                self?.startScan()
            }
        }
    }

    nonisolated func centralManager(
        _ central: CBCentralManager,
        didDiscover peripheral: CBPeripheral,
        advertisementData: [String: Any],
        rssi RSSI: NSNumber
    ) {
        let name = peripheral.name ?? advertisementData[CBAdvertisementDataLocalNameKey] as? String
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            guard KeyfobBLE.matchesName(name), let name else { return }
            if self.DEVICE_SEARCH_ID == "NOT_SET" {
                if self.discoveredNames.insert(name).inserted {
                    Bridge.sendDiscoveredDevice(
                        ControllerTypes.KEYFOB,
                        name,
                        deviceAddress: peripheral.identifier.uuidString,
                        rssi: RSSI.intValue
                    )
                }
                return
            }
            if self.peripheral != nil { return }
            let target = self.DEVICE_SEARCH_ID
            let matches =
                target.isEmpty
                || name.caseInsensitiveCompare(target) == .orderedSame
                || name.localizedCaseInsensitiveContains(target)
                || peripheral.identifier.uuidString.caseInsensitiveCompare(target) == .orderedSame
            guard matches else { return }
            self.connect(peripheral)
        }
    }

    nonisolated func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        DispatchQueue.main.async { [weak self] in
            Bridge.log("Keyfob: Connected to \(peripheral.name ?? "?")")
            peripheral.discoverServices([KeyfobBLE.SERVICE_UUID])
        }
    }

    nonisolated func centralManager(
        _ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?
    ) {
        DispatchQueue.main.async { [weak self] in
            Bridge.log("Keyfob: failed to connect: \(error?.localizedDescription ?? "unknown")")
            self?.resetControllerState()
            self?.peripheral = nil
        }
    }

    nonisolated func centralManager(
        _ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?
    ) {
        DispatchQueue.main.async { [weak self] in
            Bridge.log("Keyfob: disconnected \(error?.localizedDescription ?? "")")
            self?.resetControllerState()
            self?.peripheral = nil
            self?.ctrlChar = nil
            self?.evtChar = nil
        }
    }
}

extension Keyfob: CBPeripheralDelegate {
    nonisolated func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        DispatchQueue.main.async {
            if let error {
                Bridge.log("Keyfob: service discovery error: \(error.localizedDescription)")
                return
            }
            peripheral.services?.forEach { peripheral.discoverCharacteristics(nil, for: $0) }
        }
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?
    ) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if let error {
                Bridge.log("Keyfob: characteristic discovery error: \(error.localizedDescription)")
                return
            }
            for char in service.characteristics ?? [] {
                switch char.uuid {
                case KeyfobBLE.CTRL_UUID:
                    self.ctrlChar = char
                case KeyfobBLE.EVT_UUID:
                    self.evtChar = char
                    peripheral.setNotifyValue(true, for: char)
                default:
                    break
                }
            }
        }
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral, didUpdateNotificationStateFor characteristic: CBCharacteristic, error: Error?
    ) {
        DispatchQueue.main.async { [weak self] in
            if let error {
                Bridge.log("Keyfob: notify error: \(error.localizedDescription)")
                return
            }
            if characteristic.uuid == KeyfobBLE.EVT_UUID {
                self?.markReady()
            }
        }
    }

    nonisolated func peripheral(
        _: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error _: Error?
    ) {
        let data = characteristic.value
        let uuid = characteristic.uuid
        DispatchQueue.main.async { [weak self] in
            guard let self, uuid == KeyfobBLE.EVT_UUID, let decoded = KeyfobBLE.decode(data) else { return }
            let (opcode, _, payload) = decoded
            switch opcode {
            case KeyfobBLE.EVT_BATTERY:
                if let percent = payload.first {
                    DeviceStore.shared.apply("glasses", "controllerBatteryLevel", Int(percent))
                }
            case KeyfobBLE.EVT_READY:
                self.markReady()
            case KeyfobBLE.EVT_GESTURE:
                guard let first = payload.first, let name = KeyfobBLE.gestureName(first) else { return }
                let source = payload.count > 1 ? Int32(payload[1]) : nil
                Bridge.sendTouchEvent(
                    deviceModel: ControllerTypes.KEYFOB,
                    gestureName: name,
                    timestamp: Int64(Date().timeIntervalSince1970 * 1000),
                    source: source
                )
            default:
                break
            }
        }
    }
}
